# Load CUAD dataset

Loads the official CUAD train/test JSON files and flattens them into a clause-level DataFrame
(one row per annotated span) with: contract ID, category, clause text, and span offsets.

Data files are in `data/cuad/` (see `data/cuad/README.md`). We use the official
`train_separate_questions.json` (408 contracts) and `test.json` (102 contracts) splits as-is.

In [1]:
import json
import re
from pathlib import Path

import pandas as pd

DATA_DIR = Path("../data/cuad")
TRAIN_PATH = DATA_DIR / "train_separate_questions.json"
TEST_PATH = DATA_DIR / "test.json"

## Flatten the JSON

Each file is SQuAD-style: `data[] -> paragraphs[] -> qas[] -> answers[]`. Every contract has one paragraph
whose `context` is the full contract text, and one question per category.

Note: in `train_separate_questions.json` each answer span is split out into its own question, so the
question ids end in `_0`, `_1`, ... (e.g. `...__Parties_3`). We strip that suffix to recover the category name.
`test.json` has one question per category with multiple answers instead.

In [2]:
def category_from_id(qid: str) -> str:
    """'<contract title>__Parties_3' -> 'Parties'"""
    return re.sub(r"_\d+$", "", qid.split("__")[-1])


def load_cuad(path):
    """Return (clauses_df, contracts_df).

    clauses_df:   one row per annotated span
    contracts_df: one row per contract with the full text
    """
    with open(path) as f:
        data = json.load(f)["data"]

    clause_rows, contract_rows = [], []
    for contract in data:
        contract_id = contract["title"]
        for para in contract["paragraphs"]:
            contract_rows.append({"contract_id": contract_id, "text": para["context"]})
            for qa in para["qas"]:
                category = category_from_id(qa["id"])
                for ans in qa["answers"]:
                    start = ans["answer_start"]
                    clause_rows.append({
                        "contract_id": contract_id,
                        "category": category,
                        "clause_text": ans["text"],
                        "start": start,
                        "end": start + len(ans["text"]),
                    })

    return pd.DataFrame(clause_rows), pd.DataFrame(contract_rows)

In [3]:
train_clauses, train_contracts = load_cuad(TRAIN_PATH)
test_clauses, test_contracts = load_cuad(TEST_PATH)

print(f"train: {len(train_contracts)} contracts, {len(train_clauses)} clause spans")
print(f"test:  {len(test_contracts)} contracts, {len(test_clauses)} clause spans")

train: 408 contracts, 11180 clause spans
test:  102 contracts, 2643 clause spans


In [4]:
train_clauses.head(10)

,contract_id,category,clause_text,start,end
0,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Document Name,DISTRIBUTOR AGREEMENT,44,65
1,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Parties,Distributor,244,255
2,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Parties,Electric City of Illinois L.L.C.,49574,49606
3,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Parties,Electric City of Illinois LLC,212,241
4,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Parties,Company,197,204
5,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Parties,Electric City Corp.,148,167
6,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Agreement Date,"7th day of September, 1999.",263,290
7,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Effective Date,Unless earlier terminated otherwise prov...,31058,31270
8,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Effective Date,The term of this Agreement shall be ten (10)...,5268,5541
9,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,Expiration Date,The term of this Agreement shall be ten (10)...,5268,5541


In [5]:
train_contracts.head()

,contract_id,text
0,LIMEENERGYCO_09_09_1999-EX-10-DISTRIBUTOR AGRE...,EXHIBIT 10.6\n\n ...
1,"WHITESMOKE,INC_11_08_2011-EX-10.26-PROMOTION A...",Exhibit 10.26 CONFIDENTIAL TREATMENT HAS BE...
2,NELNETINC_04_08_2020-EX-1-JOINT FILING AGREEMENT,Exhibit 1\n\nJOINT FILING AGREEMENT\n\nThe und...
3,ADAMSGOLFINC_03_21_2005-EX-10.17-ENDORSEMENT A...,REDACTED COPY\n\nCONFIDENTIAL TREATMENT REQUES...
4,"KIROMICBIOPHARMA,INC_05_11_2020-EX-10.23-CONSU...",Exhibit 10.23 Corporate Address Fannin South P...


## Sanity checks

- 41 categories in train. Test only has 40: `Price Restrictions` has no positive spans in any test contract
  (it is a rare category, ~4% of train contracts).
- Span offsets point at the clause text in the contract

In [6]:
train_cats = set(train_clauses["category"])
test_cats = set(test_clauses["category"])
print("train categories:", len(train_cats))
print("test categories: ", len(test_cats))
print("in train but not test:", train_cats - test_cats)
assert len(train_cats) == 41 and test_cats <= train_cats

train categories: 41
test categories:  40
in train but not test: {'Price Restrictions'}


In [7]:
def check_offsets(clauses, contracts):
    text_by_id = contracts.set_index("contract_id")["text"]
    sliced = [
        text_by_id[row.contract_id][row.start:row.end]
        for row in clauses.itertuples()
    ]
    return (pd.Series(sliced) == clauses["clause_text"].values).mean()

print("train offset match rate:", check_offsets(train_clauses, train_contracts))
print("test offset match rate: ", check_offsets(test_clauses, test_contracts))

train offset match rate: 1.0
test offset match rate:  1.0


## Contract-level label matrix

A multi-hot `contracts x 41 categories` matrix (1 if the contract has at least one span for that category).
Useful for the contract-level split (#8) and the EDA on class imbalance (#9).

In [8]:
def label_matrix(clauses, contracts):
    mat = pd.crosstab(clauses["contract_id"], clauses["category"]).clip(upper=1)
    # keep every contract, even ones with no spans, and a fixed column order
    return mat.reindex(index=contracts["contract_id"], columns=sorted(mat.columns), fill_value=0)

train_labels = label_matrix(train_clauses, train_contracts)
test_labels = label_matrix(test_clauses, test_contracts)

print(train_labels.shape, test_labels.shape)
train_labels.sum().sort_values(ascending=False)

(408, 41) (102, 40)


category
Document Name                         408
Parties                               407
Agreement Date                        377
Governing Law                         354
Expiration Date                       335
Effective Date                        320
Anti-Assignment                       302
Cap On Liability                      231
License Grant                         205
Audit Rights                          176
Termination For Convenience           154
Post-Termination Services             153
Renewal Term                          150
Exclusivity                           147
Insurance                             134
Minimum Commitment                    133
Revenue/Profit Sharing                131
Non-Transferable License              116
Ip Ownership Assignment               101
Uncapped Liability                     98
Non-Compete                            96
Change Of Control                      95
Notice Period To Terminate Renewal     95
Covenant Not To Sue      

---

# Clean the data

Issue #7: resolve malformed annotation spans, drop duplicate contracts and clauses, and strip
formatting artifacts.

The checks below found the CUAD release is already in good shape, so this section is mostly evidence
that we looked. The one real change is dropping a duplicated contract.

## Annotation spans

The offset check above was a single pass/fail rate. Break it down: empty spans, whitespace-only spans,
offsets running past the end of the contract, and offsets whose text does not match.

In [9]:
def audit_spans(clauses, contracts):
    text_by_id = contracts.set_index("contract_id")["text"]
    issues = {"zero_length": 0, "whitespace_only": 0, "out_of_bounds": 0,
              "offset_mismatch": 0, "leading_trailing_ws": 0}
    for row in clauses.itertuples():
        text, span = text_by_id[row.contract_id], row.clause_text
        if len(span) == 0:
            issues["zero_length"] += 1
        elif not span.strip():
            issues["whitespace_only"] += 1
        elif span != span.strip():
            issues["leading_trailing_ws"] += 1
        if row.start < 0 or row.end > len(text):
            issues["out_of_bounds"] += 1
        elif text[row.start:row.end] != span:
            issues["offset_mismatch"] += 1
    return pd.Series(issues)


pd.DataFrame({
    "train": audit_spans(train_clauses, train_contracts),
    "test": audit_spans(test_clauses, test_contracts),
})

,train,test
zero_length,0,0
whitespace_only,0,0
out_of_bounds,0,0
offset_mismatch,0,0
leading_trailing_ws,0,0


All zero, so there is nothing to repair. The 41 categories and 13,823 spans come through intact.

## Duplicates

Contracts whose text appears twice, and any contract leaking across the train/test boundary.

In [10]:
dups = pd.concat([train_contracts.assign(split="train"),
                  test_contracts.assign(split="test")])
dups = dups[dups.duplicated(["split", "text"], keep=False)]
print("contracts shared between train and test:",
      len(set(train_contracts["text"]) & set(test_contracts["text"])))
dups[["split", "contract_id"]].assign(chars=dups["text"].str.len())

contracts shared between train and test: 0


,split,contract_id,chars
298,train,"ADUROBIOTECH,INC_06_02_2020-EX-10.7-CONSULTING...",12020
303,train,"ADUROBIOTECH,INC_06_02_2020-EX-10.7-CONSULTING...",12020


No leakage. One contract is in the release twice under slightly different titles, so drop a copy.

In [11]:
def drop_duplicate_contracts(clauses, contracts):
    """Keep the first copy of any contract whose text appears more than once."""
    keep = contracts.drop_duplicates("text", keep="first")
    dropped = set(contracts["contract_id"]) - set(keep["contract_id"])
    return clauses[~clauses["contract_id"].isin(dropped)].copy(), keep.copy(), dropped


train_clauses, train_contracts, dropped = drop_duplicate_contracts(train_clauses, train_contracts)
test_clauses, test_contracts, _ = drop_duplicate_contracts(test_clauses, test_contracts)
print("dropped:", dropped)
print("duplicate clause rows:",
      train_clauses.duplicated(["contract_id", "category", "start", "end"]).sum())
print(f"train: {len(train_contracts)} contracts / {len(train_clauses)} spans"
      f" | test: {len(test_contracts)} contracts / {len(test_clauses)} spans")

dropped: {'ADUROBIOTECH,INC_06_02_2020-EX-10.7-CONSULTING AGREEMENT'}
duplicate clause rows: 0
train: 407 contracts / 11166 spans | test: 102 contracts / 2643 spans


## Formatting artifacts

Survey what is actually in the text before deciding whether to touch it.

In [12]:
ARTIFACTS = {
    "3+ newlines": r"\n{3,}",
    "runs of 5+ spaces": r" {5,}",
    "non-ascii characters": r"[^\x00-\x7f]",
    "[*] redactions": r"\[\*+\]",
    "page markers": r"(?i)page \d+ of \d+",
    "underscore rules": r"_{5,}",
    "control characters": r"[\x00-\x08\x0b-\x1f]",
}

pd.DataFrame({
    name: {
        "train contracts affected": train_contracts["text"].str.contains(pat, regex=True).sum(),
        "test contracts affected": test_contracts["text"].str.contains(pat, regex=True).sum(),
    }
    for name, pat in ARTIFACTS.items()
}).T

,train contracts affected,test contracts affected
3+ newlines,393,99
runs of 5+ spaces,228,59
non-ascii characters,167,41
[*] redactions,87,20
page markers,28,7
underscore rules,82,23
control characters,0,0


**We leave the contract text exactly as released.** No OCR damage showed up - no control
characters, no mojibake, no broken encodings. What the survey finds is ordinary layout noise, and none of
it is worth rewriting:

- **Whitespace and unicode punctuation** (indentation runs, double spaces between words, curly quotes)
  are invisible to the models we plan to use: a transformer tokenizer collapses whitespace itself and
  TF-IDF splits on word boundaries. Normalising them would shift every character offset after the edit,
  breaking the spans that chunking (#11) relies on, for no measurable gain.
- **`[*]` redaction markers and `____` rules** sit *inside* annotated clauses, so deleting them would cut
  holes in the labelled spans.
- **`Page 3 of 12` markers** appear mid-sentence where a page broke; removing one splices two unrelated
  fragments together.

Keeping `context` byte-identical to the CUAD release also keeps our numbers comparable to the published
paper. Worth confirming with the advisor.

## Save

Written to `data/processed/` as CSV so it is easy for anyone on the team to open. That folder is
gitignored, so teammates re-run this notebook rather than pulling generated files, and the raw CUAD
release stays the only data in version control.

In [13]:
OUT_DIR = Path("../data/processed")
OUT_DIR.mkdir(exist_ok=True)

# rebuild the label matrix now that the duplicate contract is gone
train_labels = label_matrix(train_clauses, train_contracts)
test_labels = label_matrix(test_clauses, test_contracts)

for name, df in [
    ("train_clauses", train_clauses), ("train_contracts", train_contracts),
    ("test_clauses", test_clauses), ("test_contracts", test_contracts),
    ("train_labels", train_labels.reset_index()), ("test_labels", test_labels.reset_index()),
]:
    path = OUT_DIR / f"{name}.csv"
    df.to_csv(path, index=False)
    print(f"{path.name:24s} {len(df):6,} rows")

train_clauses.csv        11,166 rows


train_contracts.csv         407 rows
test_clauses.csv          2,643 rows
test_contracts.csv          102 rows
train_labels.csv            407 rows
test_labels.csv             102 rows


## Summary

| Check | Result |
| --- | --- |
| Categories | 41 in train, 40 in test (`Price Restrictions` has no test positives) |
| Malformed spans (empty, whitespace-only, out-of-bounds, mismatched) | none found in 13,823 spans |
| Duplicate contracts | 1 dropped from train (`ADUROBIOTECH ... CONSULTING AGREEMENT`) |
| Duplicate clause rows | none |
| Train/test leakage | none |
| Formatting artifacts | surveyed; text left byte-identical to the release (see reasoning above) |

Final: 407 train contracts / 11,166 spans, 102 test contracts / 2,643 spans.